core/src/test/scala/org/apache/spark/scheduler/BlacklistTrackerSuite.scala

```diff
@@ -0,0 +1,503 @@
+    assertEquivalentToSet(blacklistTracker.isNodeBlacklisted(_), Set())
+    assertEquivalentToSet(blacklistTracker.isExecutorBlacklisted(_), Set())
+
+    // when the stage completes successfully, now there is sufficient evidence we've got
+    // bad executors and node
+    blacklistTracker.updateBlacklistForSuccessfulTaskSet(0, 0, tsm.execToFailures)
+    assert(blacklistTracker.nodeBlacklist() === Set("hostA"))
+    assertEquivalentToSet(blacklistTracker.isNodeBlacklisted(_), Set("hostA"))
+    assertEquivalentToSet(blacklistTracker.isExecutorBlacklisted(_), Set("1", "2"))
+
+    clock.advance(blacklistTracker.BLACKLIST_TIMEOUT_MILLIS + 1)
+    blacklistTracker.applyBlacklistTimeout()
+    assert(blacklistTracker.nodeBlacklist() === Set())
+    assertEquivalentToSet(blacklistTracker.isNodeBlacklisted(_), Set())
+    assertEquivalentToSet(blacklistTracker.isExecutorBlacklisted(_), Set())
+  }
+
+  def trackerFixture: (BlacklistTracker, TaskSchedulerImpl) = {
+    trackerFixture()
+  }
+
+  def trackerFixture(confs: (String, String)*): (BlacklistTracker, TaskSchedulerImpl) = {
+    val conf = new SparkConf().setAppName("test").setMaster("local")
+      .set(config.BLACKLIST_ENABLED.key, "true")
+    confs.foreach { case (k, v) => conf.set(k, v) }
+    val scheduler = mockTaskSchedWithConf(conf)
+
+    clock.setTime(0)
+    blacklistTracker = new BlacklistTracker(conf, clock)
+    (blacklistTracker, scheduler)
+  }
+
+  test("executors can be blacklisted with only a few failures per stage") {
+    val (tracker, scheduler) = trackerFixture
+    // for 4 different stages, executor 1 fails a task, then executor 2 succeeds the task,
+    // and then the task set is done.  Not enough failures to blacklist the executor *within*
+    // any particular taskset, but we still blacklist the executor overall eventually
+    (0 until 4).foreach { stage =>
+      val taskSet = FakeTask.createTaskSet(1)
+      val tsm = new TaskSetManager(scheduler, Some(tracker), taskSet, 4, clock)
+      tsm.updateBlacklistForFailedTask("hostA", "1", 0)
+      tracker.updateBlacklistForSuccessfulTaskSet(stage, 0, tsm.execToFailures)
+    }
+    assertEquivalentToSet(tracker.isExecutorBlacklisted(_), Set("1"))
+  }
+
+  // if an executor has many task failures, but the task set ends up failing, don't count it
+  // against the executor
+  test("executors aren't blacklisted if task sets fail") {
+    val (tracker, scheduler) = trackerFixture
+    // for 4 different stages, executor 1 fails a task, and then the taskSet fails.
+    (0 until 4).foreach { stage =>
+      val taskSet = FakeTask.createTaskSet(1)
+      val tsm = new TaskSetManager(scheduler, Some(tracker), taskSet, 4, clock)
+      tsm.updateBlacklistForFailedTask("hostA", "1", 0)
+    }
+    assertEquivalentToSet(tracker.isExecutorBlacklisted(_), Set())
+  }
+
+  Seq(true, false).foreach { succeedTaskSet =>
+    test(s"stage blacklist updates correctly on stage completion ($succeedTaskSet)") {
```

Comment:
not totally necessary, but they also aren't completely redundant.  The previous test has multiple task sets fail -- this one has one taskset fail, but with lots of tasks in that task set.

there is some overlap, so I could eliminate the earlier one ... or going the other way, there are lots more permutations I could test as well ...